quic/s2n-quic-crypto/src/cipher_suite/ring.rs

```diff
@@ -1,84 +1,130 @@
-                    unimplemented!();
-                }
+macro_rules! key {
+    ($name:ident, $ring_cipher:path, $key_size:expr, $tag_len:expr) => {
+        pub mod $name {
+            use super::super::$name::{KEY_LEN, NONCE_LEN, TAG_LEN};
+            use crate::ring_aead::{self as aead};
+            use s2n_quic_core::crypto::scatter;
+            use zeroize::Zeroize;
+
+            pub struct Key {
+                key: aead::LessSafeKey,
             }
 
-            impl Zeroize for Key {
-                fn zeroize(&mut self) {
-                    // ring doesn't provide a way to zeroize keys currently
-                    // https://github.com/briansmith/ring/issues/15
+            impl Key {
+                pub fn new(secret: &[u8; KEY_LEN]) -> Self {
+                    let unbound_key =
+                        aead::UnboundKey::new(&$ring_cipher, secret).expect("key size verified");
+                    let key = aead::LessSafeKey::new(unbound_key);
+                    Self { key }
                 }
             }
 
-            impl crate::aead::Aead for Key {
-                type Nonce = [u8; NONCE_LEN];
-                type Tag = [u8; TAG_LEN];
-
-                #[inline]
-                fn encrypt(
-                    &self,
-                    nonce: &[u8; NONCE_LEN],
-                    aad: &[u8],
-                    payload: &mut scatter::Buffer,
-                ) -> crate::aead::Result {
-                    self.key.encrypt(nonce, aad, payload)
-                }
+            key_impl!($name, $ring_cipher, $key_size, $tag_len);
+        }
+    };
+}
 
-                #[inline]
-                fn decrypt(
-                    &self,
-                    nonce: &[u8; NONCE_LEN],
-                    aad: &[u8],
-                    input: &mut [u8],
-                    tag: &[u8; TAG_LEN],
-                ) -> crate::aead::Result {
-                    self.key.decrypt(nonce, aad, input, tag)
-                }
+macro_rules! key_impl {
+    ($name:ident, $ring_cipher:path, $key_size:expr, $tag_len:expr) => {
+        impl Key {
+            #[inline]
+            #[allow(dead_code)] // this is to maintain compatibility between implementations
+            pub fn should_update_pmtu(&self, _mtu: u16) -> bool {
+                // ring doesn't implement precomputed tables
+                false
+            }
+
+            #[inline]
+            #[allow(dead_code)] // this is to maintain compatibility between implementations
+            pub fn update(&self, secret: &[u8; KEY_LEN]) -> Self {
+                // no state to persist
+                Self::new(secret)
+            }
+
+            #[inline]
+            #[allow(dead_code)] // this is to maintain compatibility between implementations
+            pub fn update_pmtu(&mut self, _secret: &[u8; KEY_LEN], _mtu: u16) {
+                unimplemented!();
+            }
```

Comment:
I think this probably could be removed now that the custom AES code is gone. Maybe a follow up task though, since I think there are a few places that MTU isn't needed anymore

Reply:
Yep, also created an issue so we don't forget: https://github.com/aws/s2n-quic/issues/2204